### ISSUE TITLE
Feature: Warn when uploading over newer file

### ISSUE THREAD

**Issue by evtimov** [satiromarra/vscode-sftp#27]
Hello,

Apologies if this feature is already available but I can't find it.

Is it possible to have a feature to "Warn when uploading over newer file" please? So if a remote file is newer than the local working copy, some sort of a prompt maybe to ask if I want to upload my version and replace the remote one or cancel the upload, skip file etc? This would be of great help if works when "uploadOnSave" is true.

Please have a look at the following references:

![1](https://user-images.githubusercontent.com/47464444/230794771-edeaaf42-265d-4b05-8a19-b5388d9c5391.png)

![2](https://user-images.githubusercontent.com/47464444/230794774-3f03cae7-56df-4683-a158-fbe924cbdb06.png)

Thank you.

**Comment by satiromarra** [comment]
@evtimov It would be possible to add it. I can think of certain cases in which that new functionality can cause problems, anyway I will try to do something.

**Comment by satiromarra** [comment]
@evtimov  After investigating and performing some tests I have found a small difficulty, the "UploadOnSave" action is executed after saving the file so the date of the local file will always be later than the date of the destination file.

**Comment by evtimov** [comment]
@satiromarra yes of course, you are right and it totally makes sense!
I have seen this functionality in PhpStorm (as per the screenshot in the first post) and in Adobe Dreamweaver (the second screenshot). 
Maybe here you can find something that can help:  _https://youtrack.jetbrains.com/issue/WI-4825_
I believe the Timestamp on the server is used somehow...
I have also found this as request and a small discussion here: _https://github.com/liximomo/vscode-sftp/issues/215_

If we pay attention to the actual message,

> test1.php has changed on the remote server (09/04/2023 23:19) 
> since your last get or put operation (09/04/2023 23:18). 
> Putting the file may overwrite changes to the file. Do you wish to put the file anyway?

It seems that the software keeps the timestamps of the files since the get or put operation?
I would really like to be more helpful.

I have also found this explanation here: https://github.com/Natizyskunk/vscode-sftp/issues/233 

```
**Is this a similar or duplicate feature request?**
The same feature request was discussed in the original SFTP extension's repo https://github.com/liximomo/vscode-sftp/issues/215

**Is your feature request related to a problem? Please describe.**
If I have the project downloaded locally, make some changes and save it, the file will automatically upload to the remote without checking if my local file was the same as the remote. This workflow can lead to unexpected overrides if working in a team.

**Describe the solution you'd like**
Warn when uploading over newer files (add diffOnUpload setting in SFTP config). It would function as follows:

1. On file save event download matching file from the remote.
2. Compare it with the local file before editing started.
3. If it isn't different, save the file and upload it.
4. If there is a difference, cancel the save and show an alert (e.g "file on the remote is newer" [override anyway] [compare changes]).
5. If compare changes is pressed show the diff panel between the local and the remote file.

**Describe alternatives you've considered**
"ftp-kr" extension or "PHPstorm" does this already. With VSCode we could probably achieve this by creating tasks.json.

**Does this project help you?**
I use this extension on daily basis and it eases my workflow a lot. I would love to see this functionality added and believe many others will appreciate it too. Thank you for your consideration.
```

**Comment by satiromarra** [comment]
@evtimov Can you try this version? [v1.17.11-beta2](https://github.com/satiromarra/vscode-sftp/releases/download/v1.17.11-beta2/code-sftp-1.17.11-beta2.vsix)

Remember that it is a beta, I have not yet managed to solve the difference in hours between a local server and the remote server.

**Comment by evtimov** [comment]
OMG splendid work @satiromarra 

Yes I can confirm it works. One thing I noticed - If I click on Upload anyway, file uploads but when I make new local change and CTRL+S (+upload) again, it shows the same prompt again asking the same question. I would expect once I have uploaded my version of the file, on second save no further diff related questions should be asked unless file on the server has some other changes than my own.

I relate the above as a bug because if I download the remote file and make a new change, CTRL+S (+upload)  uploads my work without prompting so everything works great in that case.

There is also a case when prompt appears after the 1st save and upload, when continuing working on the same file,  pressing CTRL+S (+upload) brings back the prompt with no need as I am sure the file I already uploaded after the 1st save is the same.

Thank you so much!

![1111](https://user-images.githubusercontent.com/47464444/231786416-d228d052-2c66-400d-9c17-2f0d9baf73da.jpg)

![222](https://user-images.githubusercontent.com/47464444/231790065-12f86ee1-f73c-427a-b789-b8092cd46ed4.jpg)

**Comment by satiromarra** [comment]
@evtimov great news! 😉 Could you tell me if the server date is correct or if it has any difference?

**Comment by evtimov** [comment]
Ah yes, there is a difference, the server time-zone is set to **UTC+2** while my workstation works in **UTC+0**.

**Comment by satiromarra** [comment]
@evtimov As it will fill me time to solve the time difference, if you want you can reinstall the production version so that it is not continuously showing the message of differences.

**Comment by satiromarra** [comment]
@evtimov I have not been able to calculate (automatically) the time difference between the remote machine and the local machine, so another way is to set the difference in `sftp.json` whit the option `remoteTimeOffsetInHours`, if the server has -2 hours, set `remoteTimeOffsetInHours: -2`.

Try [beta3](https://github.com/satiromarra/vscode-sftp/releases/download/v1.17.11-beta3/code-sftp-1.17.11-beta3.vsix) and let me know.

**Comment by evtimov** [comment]
Thank you @satiromarra!
Well, in my case the remote time offset is UTC+2.
I can confirm setting  ` "remoteTimeOffsetInHours": 2` in `sftp.json` works flawlessly.
**¡Muchas gracias!** 

P.S. In case you find this thread useful: https://stackoverflow.com/questions/3245381/how-to-get-time-zone-setting-of-ftp-server

> There is no way defined in the FTP standard to determine remote server's time zone.
> If you have write permissions to the FTP server I guess you could upload the file and then calculate the difference between the file time reported by FTP and locally.

**Comment by satiromarra** [comment]
I have tried several ways to get the date from the ftp server, there are different commands but not all servers support it, the only valid way I found was the same one, when opening the connection create a file and check the date of the new file with the local time. But there are differences:

The command "LIST" is like "ls -la" and the command "MLSD" is like "modtime", "LIST" or "LIST path" Returns the entire list of files and folders with dates but without timezone, "MLSD" returns timezone but is not supported by some servers and does not work on windows servers.
(I won't go into too much detail so as not to bore you.)

First server (diferent date):
```
ftp> ls -la
....
-rw-r--r--    1 root   root            0 Apr 13 21:33 test.php

ftp> modtime test.php
test.php        04/13/2023 20:33:34 GMT
```

Second server (same date):
```
ftp> ls -la
....
-rw-r--r--    1 root   root            0 Apr 13 20:42 test.php

ftp> modtime test.php
test.php        04/13/2023 20:42:13 GMT
```

The first server returns different dates for the same file.
The second server returns the same dates for the same file.

Maybe the only way for now is to set in the configuration the difference in hours, it is not elegant but it works.

**Comment by evtimov** [comment]
`Maybe the only way for now is to set in the configuration the difference in hours, it is not elegant but it works.`

 ٩(๑･ิᴗ･ิ)۶ 🍺🍺 ٩(･ิᴗ･ิ๑)۶

**Comment by satiromarra** [comment]
I just published the new version 1.17.11

**Comment by evtimov** [comment]
This is great, thank you very much!

**Comment by andreysamode** [comment]
This looks wonderful. Unfortunately, I'm having trouble getting it to work.
I installed the v1.17.11-beta3
The "date" on my computer is Sat Jun 24 14:05:19 HST 2023
The "date" on the server is Sat Jun 24 20:05:34 EDT 2023
... so I added the following to sftp.json:
"remoteTimeOffsetInHours": 6

I then edited the file remotely and saved it.
But when I go to upload the file from my computer in VSCode, it just uploads without giving me any messages.

What am I missing?

Thanks!

**Comment by satiromarra** [comment]
@mrandrey In the extension settings, is the option to check file dates enabled?

Enable verbose mode in the settings and check the sftp output tab.

**Comment by andreysamode** [comment]
I'm not seeing the option to check file dates in the settings:
<img width="694" alt="Screen Shot 2023-06-25 at 6 31 58 AM" src="https://github.com/satiromarra/vscode-sftp/assets/444565/fbf3998e-4603-4056-a56e-1cc86f6fd329">

Here's the version I have:
<img width="404" alt="Screen Shot 2023-06-25 at 6 34 29 AM" src="https://github.com/satiromarra/vscode-sftp/assets/444565/c07e6106-edc3-4753-a7a6-93c07a931957">

Should I try a different one?

**Comment by satiromarra** [comment]
Update to the latest version

**Comment by andreysamode** [comment]
Updated to 1.18.2-beta. The following option now shows up:
<img width="676" alt="Screen Shot 2023-06-25 at 10 08 14 AM" src="https://github.com/satiromarra/vscode-sftp/assets/444565/fe6fe1d1-239d-48c0-bc97-f8f15fdac5e0">

In sftp.json, I have:
  "uploadOnSave": false,
  "useTempFile": false,
  "openSsh": false,
  "remoteTimeOffsetInHours": 6

When I upload with keyboard shortcut for sftp.upload.activeFile, the file just uploads, even if the server has newer file.
I tried setting uploadOnSave to true, and upon saving, the same thing happens.
Then I tried setting "remoteTimeOffsetInHours" to 10 and -10, and it just overrides it just the same.

I added these settings to settings.json:

"sftp.downloadWhenOpenInRemoteExplorer": true,
"sftp.printVerboseLog": true,
"sftp.debug": true,
"sftp.warnWhenUploadingOverNewerFile": true

And when I upload, the only line in output:sftp is this:
[2023-06-25 14:19:22] [info] local ➞ remote /Users/x/Web/x.x/x.php

I have the file on the server open in Vim and when I re-save it after the upload, it says that it's been changed, so the file is the correct one.

**Comment by satiromarra** [comment]
With this configuration it should work
"uploadOnSave": true,
"remoteTimeOffsetInHours": 6

With the debug or info options activated, the dates should appear in the output.

**Comment by andreysamode** [comment]
When uploadOnSave: true, I get the following in the output:
[2023-06-26 07:13:36] [info] The remote file (12/31/1969, 2:00:00 PM) vs the local file (12/31/1969, 2:00:00 PM).

**Comment by satiromarra** [comment]
Wow, It seems that the date obtained from the ftp server is not processed correctly. Do you know which ftp server it is?

**Comment by andreysamode** [comment]
It's just regular sftp as far as I can tell. Same login as ssh, used with a key. I can run a command on the server to get more info if that would be helpful. Both local and remote have the date obliterated, so I don't think it's the server. I'm on a mac, uploading to CentOS. Nothing fancy as far as ssh/sftp configurations.

**Comment by satiromarra** [comment]
@mrandrey  Try this beta [v1.18.2-beta2](https://github.com/satiromarra/vscode-sftp/releases/download/v1.18.2-beta2/code-sftp-1.18.2-beta2.vsix)
Enable the verbose option, shows the Output tab when save a file. A message in "json" format should appear with the file data, dates, etc.
The folder of project is a mapped folder or contained in a mapped folder?

**Comment by andreysamode** [comment]
New version installed, restarted, all the verbose/debug/uploadOnSave settings enabled.

It worked with the file initially:

<img width="469" alt="Screen Shot 2023-06-26 at 9 52 26 AM" src="https://github.com/satiromarra/vscode-sftp/assets/444565/981dbb0b-8226-4f49-bf57-0fc93a50f93b">

... and when I updated the file again on the server, it's overriding it again without any prompts.

Here's the log:
[2023-06-26 09:56:01] [debug] Outbound: Sending ping (GLOBAL_REQUEST: keepalive@openssh.com)
[2023-06-26 09:56:01] [debug] Inbound: Received REQUEST_FAILURE
[2023-06-26 09:56:12] [debug] Outbound: Sending CHANNEL_DATA (r:0, 53)
[2023-06-26 09:56:12] [debug] SFTP: Outbound: Buffered LSTAT
[2023-06-26 09:56:13] [debug] Inbound: CHANNEL_DATA (r:0, 41)
[2023-06-26 09:56:13] [debug] SFTP: Inbound: Received ATTRS (id:72)
[2023-06-26 09:56:13] [verbose] {"uri":"/Users/x/Web/x.info/x.php","data":{"local":{"type":2,"size":12885,"mode":420,"mtime":1687809350464,"atime":1687809367840},"remote":{"type":2,"mode":420,"size":12886,"mtime":1687787769000,"atime":1687787769000}}}
[2023-06-26 09:56:13] [info] The remote file (6/26/2023, 3:56:09 AM) vs the local file (6/26/2023, 9:55:50 AM).
[2023-06-26 09:56:13] [info] [file-save] /Users/x/Web/x.info/x.php
[2023-06-26 09:56:13] [trace] handle upload file for /Users/x/Web/x.info/x.php
[2023-06-26 09:56:13] [debug] Outbound: Sending CHANNEL_DATA (r:0, 36)
[2023-06-26 09:56:13] [debug] SFTP: Outbound: Buffered MKDIR
[2023-06-26 09:56:13] [debug] Inbound: CHANNEL_DATA (r:0, 28)
[2023-06-26 09:56:13] [debug] SFTP: Inbound: Received STATUS (id:73, 4, "Failure")
[2023-06-26 09:56:13] [debug] Outbound: Sending CHANNEL_DATA (r:0, 32)
[2023-06-26 09:56:13] [debug] SFTP: Outbound: Buffered LSTAT
[2023-06-26 09:56:14] [debug] Inbound: CHANNEL_DATA (r:0, 41)
[2023-06-26 09:56:14] [debug] SFTP: Inbound: Received ATTRS (id:74)
[2023-06-26 09:56:14] [debug] Outbound: Sending CHANNEL_DATA (r:0, 61)
[2023-06-26 09:56:14] [debug] SFTP: Outbound: Buffered OPEN
[2023-06-26 09:56:14] [debug] Inbound: CHANNEL_DATA (r:0, 17)
[2023-06-26 09:56:14] [debug] SFTP: Inbound: Received HANDLE (id:75)
[2023-06-26 09:56:14] [debug] Outbound: Sending CHANNEL_DATA (r:0, 17)
[2023-06-26 09:56:14] [debug] SFTP: Outbound: Buffered FSTAT
[2023-06-26 09:56:15] [debug] Inbound: CHANNEL_DATA (r:0, 41)
[2023-06-26 09:56:15] [debug] SFTP: Inbound: Received ATTRS (id:76)
[2023-06-26 09:56:15] [debug] Outbound: Sending CHANNEL_DATA (r:0, 25)
[2023-06-26 09:56:15] [debug] SFTP: Outbound: Buffered FSETSTAT
[2023-06-26 09:56:15] [debug] Inbound: CHANNEL_DATA (r:0, 28)
[2023-06-26 09:56:15] [debug] SFTP: Inbound: Received STATUS (id:77, 0, "Success")
[2023-06-26 09:56:15] [debug] Outbound: Sending CHANNEL_DATA (r:0, 12914)
[2023-06-26 09:56:15] [debug] SFTP: Outbound: Sent WRITE (id:78)
[2023-06-26 09:56:15] [debug] Inbound: CHANNEL_DATA (r:0, 28)
[2023-06-26 09:56:15] [debug] SFTP: Inbound: Received STATUS (id:78, 0, "Success")
[2023-06-26 09:56:15] [debug] Outbound: Sending CHANNEL_DATA (r:0, 29)
[2023-06-26 09:56:15] [debug] SFTP: Outbound: Buffered FSETSTAT
[2023-06-26 09:56:16] [debug] Inbound: CHANNEL_DATA (r:0, 28)
[2023-06-26 09:56:16] [debug] SFTP: Inbound: Received STATUS (id:79, 0, "Success")
[2023-06-26 09:56:16] [debug] Outbound: Sending CHANNEL_DATA (r:0, 17)
[2023-06-26 09:56:16] [debug] SFTP: Outbound: Buffered CLOSE
[2023-06-26 09:56:16] [debug] Inbound: CHANNEL_DATA (r:0, 28)
[2023-06-26 09:56:16] [debug] SFTP: Inbound: Received STATUS (id:80, 0, "Success")
[2023-06-26 09:56:16] [info] local ➞ remote /Users/x/Web/x.info/x.php

When I set the remoteTimeOffsetInHours to 0, it works fine. Since mtime is a unix timestamp, it's not timezone-dependent, so that makes sense.

**Comment by satiromarra** [comment]
But something strange has happened,
This message must be launched before uploading and after saving the file on the disk (in second place).
```[2023-06-26 09:45:05] [info] The remote file (Invalid Date) vs the local file (6/26/2023, 7:13:36 AM).```

This message must be launched before uploading and before saving the file to the disk (in first place).
```[2023-06-26 09:45:22] [verbose] {"uri":"/Users/x/Web/[x.info/x.php](http://x.info/x.php)","data":{"local":{"type":2,"size":11648,"mode":420,"mtime":1687808705156,"atime":1687808705184},"remote":{"type":2,"mode":420,"size":11648,"mtime":1687799616000,"atime":1687786377000}}}```

But it seems that they have been executed in reverse order and 15 seconds apart.

**Comment by andreysamode** [comment]
I ended up editing my last post, so you might have missed this:

When I set the remoteTimeOffsetInHours to 0, it works fine. Since mtime is a unix timestamp, it's not timezone-dependent, so that makes sense.

Let me know if you'd like me to try another version and run some tests. I really appreciate this functionality.

**Comment by satiromarra** [comment]
In a way "mtime" does depend on the time zone, the date of the remote and local file are not in the same time zone and when comparing them in unix format it will be incorrect, hence you have to subtract / add the difference in hours.
If the server has more than 6 hours than your local time, remoteTimeOffsetInHours must be 6.

If your local time is "6/26/2023, 9:55:50 AM" the remote time must be "6/26/2023, 3:55:50 PM"
EDT -> HST = +6 hours.

Does the connection to the server take a long time? That can be a problem.

**Comment by andreysamode** [comment]
That's the odd thing. In the log above, while remoteTimeOffsetInHours was 6:

{"type":2,"size":12885,"mode":420,"mtime":1687809350464,"atime":1687809367840},"remote":{"type":2,"mode":420,"size":12886,"mtime":1687787769000,"atime":1687787769000}}}

If I convert mtime on epochconverter.com:
local: GMT: Monday, June 26, 2023 7:55:50.464 PM
remote: GMT: Monday, June 26, 2023 1:56:09 PM

The difference is 6 hours (the time specified in remoteTimeOffsetInHours), plus the seconds it took me to do all the saving and uploading.

Which means, it's not affected by the local timezone. And it works if I set it to 0, even though I'm on a different timezone.

I always thought unix timestamps were a way to get around timezones. It's seconds since...
Here's what Google said on "is unix timestamp affected by timezone":
"Returns the current time measured in the number of seconds since the Unix Epoch (January 1 1970 00:00:00 GMT). Note: Unix timestamps do not contain any information with regards to any local timezone."

**Comment by satiromarra** [comment]
Yes, really the unix time are the seconds elapsed since "January 1 1970 00:00:00 GMT" and you always have to compare it with GMT.
The FTP server when you get the date of the file you get it as "2023-07-26 16:00:00" without indicating the time zone and in your case that time is EDT and your local time is HST, when converting both to unix you are comparing incorrect times and that is why to an mtime you have to add or subtract the difference.
It's like comparing coins, 100 is greater than 100? no, but is £100 greater than $100?.

This already means opening a discussion ;)

The date of the remote file should be equal to or less than the date of the local file.
I will have to review the process, I use it with remoteTimeOffsetInHours: 2 and I have not had problems, and I use 3 machines with (2 linux & 1 macos) different environments.

**Comment by andreysamode** [comment]
Just restarted VSCode with 0 offset. Uploaded the file, then modified it on the server, waited for 30 secs and uploaded it again. Got the override prompt:

<img width="465" alt="Screen Shot 2023-06-26 at 11 31 55 AM" src="https://github.com/satiromarra/vscode-sftp/assets/444565/c8ed48c2-3de5-4f83-bc0d-d27ed9ff49cc">

I guess whatever works? I'm happy to run more tests if you like.

Also, would it be possible to enable the override check on Right Click > Upload with uploadOnSave turned off? It doesn't seem to do the date check.

Thanks!

**Comment by evtimov** [comment]
Hi @satiromarra  when you get a chance, can you review the timezones offset. My local tme is UTC +0, GMT and my remote server is EET, UTC+2 hours  but to get this feature working i need to set ` "remoteTimeOffsetInHours": 3` (it should be **2**, right?) in sftp.json ?

It used to work fine, maybe something related to the recent daylight saving time change? :D I know that it doesnt make any sense... 

I double-checked the remote server now shows Mon Oct 30 23:06:49 EET 2023 and here in the UK is currently 21:06 (9:06 PM). 

So, to summarize, setting ` "remoteTimeOffsetInHours": 3` works great with 2 hours time difference between local and remote. I use the latest version v1.18.1

![image](https://github.com/satiromarra/vscode-sftp/assets/47464444/834d924b-5e63-43d4-8e3b-4926a25cdc37)

Thank you.
